ethcore/src/snapshot/service.rs

```diff
@@ -407,54 +417,129 @@ impl Service {
 	/// The recover flag indicates whether to recover the restored snapshot.
 	pub fn init_restore(&self, manifest: ManifestData, recover: bool) -> Result<(), Error> {
 		let rest_dir = self.restoration_dir();
-
-		let mut res = self.restoration.lock();
-
-		self.state_chunks.store(0, Ordering::SeqCst);
-		self.block_chunks.store(0, Ordering::SeqCst);
-
-		// tear down existing restoration.
-		*res = None;
+		let rest_db = self.restoration_db();
+		let recovery_temp = self.temp_recovery_dir();
+		let prev_chunks = self.prev_chunks_dir();
 
 		// delete and restore the restoration dir.
-		if let Err(e) = fs::remove_dir_all(&rest_dir) {
+		if let Err(e) = fs::remove_dir_all(&prev_chunks) {
 			match e.kind() {
 				ErrorKind::NotFound => {},
 				_ => return Err(e.into()),
 			}
 		}
 
-		fs::create_dir_all(&rest_dir)?;
+		{
+			self.ready.store(false, Ordering::SeqCst);
```

Comment:
Need to be careful with resetting the state to `false` in case of any error. IMHO better to make sure that it's only done after the lock is acquired and we know that restoration is still needed (i.e. `res != None`?)

Reply:
I updated this bit so that the restoration stays locked until the initialization is done.

Reply:
https://github.com/paritytech/parity/pull/8544/commits/0f1e5b414f74bad2b15d992c752d465b4832a21d